### PR TITLE
Move preprocessor definitions to configuration header

### PR DESCRIPTION
**Pull request by fknorr**
Celerity's list of compile-time feature flags and options has grown considerably over time, and moving them from the command line to a generated header file will keep them more manageable.

Unfortunately we already have the canonical name `config.h` in use for the run-time configuration, so instead of causing confusion with a second file of the same name, I've chosen to move them into our existing generated header `version.h`. Not perfect for discoerability, but still semantically meaningful as that file enumerates all config options for the particular version of Celerity that's being built.

This PR also replaces the somewhat ad-hoc SYCL impl detection macros with a uniform set of `CELERITY_SYCL_IS_*` defines.

Also I've made sure all of our own defines are either 0 or 1, never undefined, and added `-Wundef` - this should avoid all future confusion around `#if` vs `#if defined` .

### CODE REVIEW

File: include/version.h.in
Comment (PeterTh) [inline include/version.h.in:1]: Should this still be called `version` or would `config` be better now?

File: include/version.h.in [hunk 1]
```diff
@@ -1,5 +1,21 @@
 #pragma once
 
+#cmakedefine01 CELERITY_SYCL_IS_ACPP
+#cmakedefine01 CELERITY_SYCL_IS_DPCPP
+#cmakedefine01 CELERITY_SYCL_IS_SIMSYCL
+
+// CELERITY_DETAIL_ENABLE_DEBUG is specified on the command line
```
Comment (PeterTh) [inline include/version.h.in:7]: Why is that? Is it so that it can be changed while keeping the same Cmake build?

**Review by PeterTh — APPROVED** [review]

File: include/version.h.in
Comment (PeterTh) [reply include/version.h.in:1]: I should have read the PR message. Not ideal but understandable.

File: include/version.h.in (hunk 1 above)
Comment (fknorr) [reply include/version.h.in:7]: It's defined in terms of a generator expression to support multi-configuration builds (aka, switching between Debug and Release in MSVC).

**Review by psalz — APPROVED** [review]

**Pull request merged**